tiny-event-sourcing-kafka/src/main/kotlin/ru/quipy/kafka/core/OngoingGroupManager.kt

```diff
@@ -0,0 +1,101 @@
+
+/**
+ * [OngoingGroupManager] is responsible for managing ongoing groups of domain events.
+ * It provides methods to convert domain events to external events.
+ */
+class OngoingGroupManager(
+    private val domainGroupRegistry: DomainGroupRegistry,
+    private val ongoingGroupStorage: OngoingGroupStorage,
+    private val externalEventMapperRegistry: ExternalEventMapperRegistry,
+    private val eventMapper: EventMapper,
+    private val externalEventMapper: ExternalEventMapper
+) {
+
+    private val ongoingGroups = mutableMapOf<String, OngoingGroup<DomainEventsGroup>>()
+
+    suspend fun convertToExternalEventRecords(event: Event<*>, aggregateId: String): List<ExternalEventRecord> {
+        val domainGroup = domainGroupRegistry.getGroupFromDomainEvent(event::class)
+
+        return if (domainGroup == null) {
+            convertDomainEvent(event)
+        } else {
+            convertDomainGroup(event, domainGroup, aggregateId)
+        }
+    }
+
+    private fun convertDomainEvent(event: Event<out Aggregate>): List<ExternalEventRecord> {
+        val mapper = externalEventMapperRegistry.getOneToManyMapperFrom(event::class)
+        val integrationEventMapper = createEventMapper<DomainEventToExternalEventsMapper<Event<out Aggregate>>>(mapper)
+
+        val externalEvents = integrationEventMapper.toExternalEvents(event)
+        return externalEvents.map { createExternalEventRecord(it) }
+    }
+
+    private suspend fun convertDomainGroup(
+        event: Event<out Aggregate>,
+        domainGroup: KClass<DomainEventsGroup>,
+        aggregateId: String
+    ): List<ExternalEventRecord> {
+        ongoingGroups.putIfAbsent(
+            aggregateId,
+            OngoingGroup(
+                domainGroup, domainGroupRegistry, ongoingGroupStorage, eventMapper, aggregateId
+            )
+        )
+
+        val ongoingGroup = ongoingGroups[aggregateId]!!
+
+        ongoingGroup.addToAggregation(event)
+
+        return ongoingGroup.takeIf { it.isReadyForAggregation() }
+            ?.let {
+                val mapper = externalEventMapperRegistry.getManyToManyMapperFrom(domainGroup)
+                val integrationEventMapper = createGroupMapper<DomainGroupToExternalEventsMapper<DomainEventsGroup>>(mapper)
+
+                val externalEvents = it.mapToExternalEvent { events ->
+                    try {
+                        integrationEventMapper.toExternalEvents(events as List<DomainEventsGroup>)
+                    } catch (e: Exception) {
+                        throw e
+                    }
+                }
+
+                externalEvents.map { createExternalEventRecord(it) }
+            }
+            ?: emptyList()
+    }
+
+
+    private inline fun <reified T : Any> createEventMapper(mapperClass: KClass<out DomainEventToExternalEventsMapper<out Event<out Aggregate>>>): T {
+        return Class.forName(mapperClass.qualifiedName)
+            .getDeclaredConstructor()
+            .newInstance() as T
+    }
```

Comment:
I would say that this module should no nothing about the spring and its possibilities. 

BUT!
1. I would cache mappers cause it's too expensive to create new each time
2. I would give the client the ability to INJECT list of pre-created mappers to the manager. And then we don't care if them will be created and inject by spring (in spring config) of manually. Also it will give the client an ability to not to be restricted by only "default" constructors. If there is no suitable mapper then we can try to create by reflection and cache it

Reply:
I did this:

first, using reflection, I save mappers that are marked with a special annotation
after this, the user can specify in the configuration a list of mappers that he would like to inject using special manager methods